lib/Dialect/FIRRTL/Transforms/EmitMetadata.cpp

```diff
@@ -0,0 +1,249 @@
+    Annotation anno(annoDict);
+    auto filenameAttr = anno.getMember<StringAttr>("filename");
+    if (!filenameAttr) {
+      circuitOp->emitError(annoClass) << " requires a filename";
+      return false;
+    }
+    filename = filenameAttr.getValue();
+    // Remove the annotation from the circuit.
+    AnnotationSet::removeAnnotations(circuitOp, annoClass);
+    return true;
+  };
+
+  // Get the filenames from the annotations.
+  auto *dutBlackboxAnnoClass =
+      "sifive.enterprise.firrtl.SitestBlackBoxAnnotation";
+  auto *testBlackboxAnnoClass =
+      "sifive.enterprise.firrtl.SitestTestHarnessBlackBoxAnnotation";
+  StringRef dutFilename, testFilename;
+  if (!getFilename(dutBlackboxAnnoClass, dutFilename) ||
+      !getFilename(testBlackboxAnnoClass, testFilename))
+    return failure();
+
+  // If we don't have either annotation, no need to run this pass.
+  if (dutFilename.empty() && testFilename.empty())
+    return success();
+
+  auto *body = circuitOp.getBody();
+
+  // Find the device under test and create a set of all modules underneath it.
+  auto *dutAnnoClass = "sifive.enterprise.firrtl.MarkDUTAnnotation";
+  DenseSet<Operation *> dutModuleSet;
+  auto it = llvm::find_if(*body, [&](Operation &op) -> bool {
+    return AnnotationSet(&op).hasAnnotation(dutAnnoClass);
+  });
+  if (it != body->end()) {
+    auto instanceGraph = getAnalysis<InstanceGraph>();
+    auto *node = instanceGraph.lookup(&(*it));
+    llvm::for_each(llvm::depth_first(node), [&](InstanceGraphNode *node) {
+      dutModuleSet.insert(node->getModule());
+    });
+  }
+
+  // Any extmodule with these annotations or one of these ScalaClass classes
+  // should be excluded from the blackbox list.
+  auto *scalaClassAnnoClass = "sifive.enterprise.firrtl.ScalaClassAnnotation";
+  std::array<StringRef, 2> classBlackList = {
+      "freechips.rocketchip.util.BlackBoxedROM", "chisel3.shim.CloneModule"};
+  std::array<StringRef, 2> blackListedAnnos = {
+      "firrtl.transforms.BlackBoxInlineAnno",
+      "firrtl.transforms.BlackBoxResourceAnno"};
+
+  // Find all extmodules in the circuit. Check if they are black-listed from
+  // being included in the list. If they are not, separate them into two groups
+  // depending on if theyre in the DUT or the test harness.
+  SmallVector<StringRef> dutModules;
+  SmallVector<StringRef> testModules;
+  for (auto extModule : circuitOp.getBody()->getOps<FExtModuleOp>()) {
+    // If the module doesn't have a defname, then we can't record it properly.
+    // Just skip it.
+    if (!extModule.defname())
+      continue;
+
+    // If its a generated blackbox, skip it.
+    AnnotationSet annos(extModule);
+    if (llvm::any_of(blackListedAnnos, [&](auto blackListedAnno) {
+          return annos.hasAnnotation(blackListedAnno);
+        }))
+      continue;
+
+    // If its a blacklisted scala class, skip it.
+    auto scalaAnnoDict = annos.getAnnotation(scalaClassAnnoClass);
+    if (scalaAnnoDict) {
```

Comment:
This could be inlined:

```suggestion
    if (auto scalaAnnoDict = annos.getAnnotation(scalaClassAnnoClass)) {
```